probe/plugins/registry.go

```diff
@@ -0,0 +1,271 @@
+	watcher, err := fswatch.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	r := &Registry{
+		rootPath:          rootPath,
+		apiVersion:        apiVersion,
+		handshakeMetadata: handshakeMetadata,
+		pluginsBySocket:   map[string]*Plugin{},
+		watcher:           watcher,
+		done:              make(chan struct{}),
+	}
+	if err := r.addPath(context.Background(), r.rootPath); err != nil {
+		r.Close()
+		return nil, err
+	}
+	go r.loop()
+	return r, nil
+}
+
+// add recursively crawls the path provided, adding it to the watcher, and
+// looking for any existing sockets, loading them as plugins.
+func (r *Registry) addPath(ctx context.Context, path string) error {
+	var statT syscall.Stat_t
+	if err := fs.Stat(path, &statT); err != nil {
+		return err
+	}
+	if err := r.watcher.Add(path); err != nil {
+		return err
+	}
+	// TODO: use of fs.Stat (which is syscall.Stat) here makes this linux specific.
+	switch statT.Mode & syscall.S_IFMT {
+	case syscall.S_IFDIR:
+		files, err := fs.ReadDir(path)
+		if err != nil {
+			log.Errorf("plugins: error loading path %s: %v", path, err)
+			return nil
+		}
+		for _, file := range files {
+			if err := r.addPath(ctx, filepath.Join(path, file.Name())); err != nil {
+				log.Errorf("plugins: error loading path %s: %v", filepath.Join(path, file.Name()), err)
+				return nil
+			}
+		}
+	case syscall.S_IFSOCK:
+		if plugin, ok := r.pluginsBySocket[path]; ok {
+			log.Infof("plugins: plugin already exists %s: conflicting %s", plugin.ID, path)
+			return nil
+		}
+		tr, err := transport(path, pluginTimeout)
+		if err != nil {
+			log.Errorf("plugins: error loading plugin %s: %v", path, err)
+			return nil
+		}
+		client := &http.Client{Transport: tr, Timeout: pluginTimeout}
+		r.pluginsBySocket[path] = NewPlugin(ctx, path, client, r.apiVersion, r.handshakeMetadata)
+	default:
+		log.Infof("plugins: unknown filemode %s", path)
+	}
+	return nil
+}
+
+func (r *Registry) removePath(ctx context.Context, path string) error {
+	r.watcher.Remove(path)
+	if plugin, ok := r.pluginsBySocket[path]; ok {
+		delete(r.pluginsBySocket, path)
+		return plugin.Close()
+	}
+	return nil
+}
+
+func (r *Registry) loop() {
+	for {
+		select {
+		case <-r.done:
+			return
+		case evt := <-r.watcher.Events():
+			handlers := map[fsnotify.Op]func(context.Context, string) error{
+				fsnotify.Create: r.addPath,
+				fsnotify.Remove: r.removePath,
+				fsnotify.Chmod:  r.addPath,
+			}
+			if handler, ok := handlers[evt.Op]; ok {
+				r.lock.Lock()
+				if err := handler(context.Background(), evt.Name); err != nil {
+					log.Errorf("plugins: event %v: error: %v", evt, err)
+				}
+				r.lock.Unlock()
+			} else {
+				log.Errorf("plugins: event %v: no handler", evt)
+			}
+		case err := <-r.watcher.Errors():
+			log.Errorf("plugins: error: %v", err)
+		}
+	}
+}
+
+// ForEach walks through all the plugins running f for each one.
+func (r *Registry) ForEach(f func(p *Plugin)) {
```

Comment:
I would probably inline this since it's only used once (unless so clearly foresee more uses).